File: backend/app/schemas/check.py
"""校验记录相关 Schema"""
from datetime import datetime
from pydantic import BaseModel, Field

from app.schemas.instance import CheckRecordBrief, DetailFileBrief, InstanceNodeBrief


class CheckListItem(BaseModel):
    """校验列表项"""
    id: int
    instance_id: int
    instance_name: str
    node_id: int
    node_name: str
    task_id: int
    submitter_name: str = ""  # 节点负责人姓名
    status: str  # pending/passed/returned/terminated
    round: int = 1
    deadline: str | None = None       # 截止时间（ISO 格式）
    is_overdue: bool = False           # 是否逾期
    days_remaining: int | None = None  # 剩余天数（负数=已逾期）
    created_at: datetime | None = None

    model_config = {"from_attributes": True}


class CheckDetail(BaseModel):
    """校验详情 —— 含文件和校验进度"""
    id: int
    instance_id: int
    instance_name: str
    instance_status: str = ""
    initiator_id: int
    initiator_name: str = ""
    submitter_id: int
    submitter_name: str = ""
    priority: str = "normal"
    difficulty: str = "1"  # 难度等级（1-4）
    node_id: int
    node_name: str
    node_description: str | None = None  # 节点说明
    task_id: int
    checker_id: int
    checker_name: str = ""
    status: str
    opinion: str | None = None
    time_limit_days: int | None = None  # 完成时限（工作日）
    deadline: datetime | None = None  # 截止时间
    round: int = 1  # 当前轮次
    # 进度条
    total_nodes: int = 0
    current_node_index: int = 0
    nodes: list[InstanceNodeBrief] = []
    # 实例全部文件（展示用）
    files: list[DetailFileBrief] = []
    # 仅本节点文件（签批预览用，后端过滤）
    node_files: list[DetailFileBrief] = []
    # 负责人备注
    assignee_note: str | None = None
    # 并行校验进度
    check_progress: list[CheckRecordBrief] = []
    # 节点签批配置
    require_assignee_signature: bool = True
    require_checker_signature: bool = True
    require_approver_signature: bool = True
    signature_x: float = 400
    signature_y: float = 100
    signature_page: int = -1
    # 当前校验人的签名图片 URL
    current_signature_url: str | None = None
    # 角色维度签名默认配置（从 SystemConfig 读取）
    role_signature: dict | None = None
    decided_at: datetime | None = None
    created_at: datetime | None = None


class CheckAction(BaseModel):
    """校验操作（通过/退回）"""
    opinion: str | None = Field(None, max_length=500, description="校验意见（通过时可选，退回时必填）")
    # 签批：支持多文档多签名
    signatures: list[dict] | None = Field(None, description="签名列表 [{file_id, signature_x, signature_y, signature_page}]")

File: frontend/src/api/check.ts
/** 校验 API */
import request from './request'
import type { PaginatedResponse } from './index'
import type { SignatureSlot } from './signature'

export interface CheckListItem {
  id: number
  instance_id: number
  instance_name: string
  node_id: number
  node_name: string
  task_id: number
  submitter_name: string
  status: string
  round: number
  deadline: string | null       // 截止时间（ISO 格式）
  is_overdue: boolean            // 是否逾期
  days_remaining: number | null  // 剩余天数（负数=已逾期）
  created_at: string | null
}

export interface CheckDetail {
  id: number
  instance_id: number
  instance_name: string
  instance_status: string
  initiator_id: number
  initiator_name: string
  submitter_id: number
  submitter_name: string
  priority: string
  difficulty: string  // 难度等级（1-4）
  node_id: number
  node_name: string
  node_description: string | null  // 节点说明
  task_id: number
  checker_id: number
  checker_name: string
  status: string
  opinion: string | null
  time_limit_days: number | null  // 完成时限（工作日）
  deadline: string | null         // 截止时间
  round: number                   // 当前轮次
  total_nodes: number
  current_node_index: number
  nodes: { id: number; name: string; is_start: boolean; is_end: boolean; status: string; sort_order: number }[]
  files: { id: number; original_name: string; mime_type: string | null; file_size: number | null; uploader_name: string; upload_type: string; round: number; node_id: number | null; node_name: string; created_at: string | null }[]
  node_files: { id: number; original_name: string; mime_type: string | null; file_size: number | null; round: number }[]  // 仅本节点文件（签批用，后端过滤）
  assignee_note: string | null
  check_progress: { id: number; checker_id: number; checker_name: string; status: string; opinion: string | null; round: number; decided_at: string | null }[]
  // 节点签批配置
  require_assignee_signature: boolean
  require_checker_signature: boolean
  require_approver_signature: boolean
  signature_x: number
  signature_y: number
  signature_page: number
  current_signature_url: string | null
  role_signature: { x: number; y: number } | null  // 角色维度签名默认坐标
  decided_at: string | null
  created_at: string | null
}

export async function getChecks(params: { status?: string; keyword?: string; page?: number; page_size?: number }) {
  const res = await request.get('/checks', { params })
  return res.data as PaginatedResponse<CheckListItem>
}

export async function getCheckDetail(id: number): Promise<CheckDetail> {
  const res = await request.get(`/checks/${id}`)
  return res.data
}

/** 校验通过 —— 支持签名 */
export async function passCheck(id: number, opinion?: string | null, signatures?: SignatureSlot[] | null) {
  const res = await request.post(`/checks/${id}/pass`, { opinion, signatures })
  return res.data
}

export async function returnCheck(id: number, opinion: string) {
  const res = await request.post(`/checks/${id}/return`, { opinion })
  return res.data
}
